### DISCUSSION TITLE
registerValidateMessageTemplateEnigne更详细的文档说明？

### DISCUSSION THREAD

**Discussion by betterRunner** [alibaba/formily#1652] [Q&A 2.x] [answered]
[文档](https://core.formilyjs.org/api/entry/form-validator-registry#registervalidatemessagetemplateenigne)目前只描述了注册的细节但是没有描述'x-validator'配合写法的例子，[这里](https://v2.formilyjs.org/zh-CN/guide/advanced/validate)的例子好像也跳过了，能否稍微给予更详细的使用说明？感谢～

**Comment by janryWang** [comment, marked as answer]
这种是非常底层的API，90%以上用户用不到，用到的也是非常资深的用户了，直接看源码即可

**Reply by betterRunner** [reply]
了解，但是我感觉我的需求是相对比较常见的：其实我只是需要全局修改reqired的样式，默认的是“该字段是必填字段”，我希望改成“请输入xxx”这种动态的文案，看了validate章节的几种方法好像只有`registerValidateMessageTemplateEnigne`配合`registerValidateLocale`实现。（用registerValidateRules覆盖原有的require规则会丢失必填的星星符号）。或者其实有更简单的解决方案？十分感谢！

**Reply by janryWang** [reply]
修改文案，直接registerValidateLocales

**Reply by betterRunner** [reply]
额，registerValidateLocales好像只可以固定文案是吗，我是需要自定义的文案“请输入xxx”，xxx是每个字段都不一样的动态改变

**Reply by janryWang** [reply]
这个不好搞的，校验器是感知不到字段模型的，也就是你拿不到title来拼装

**Reply by janryWang** [reply]
除非，你在effects 里手动设置errors

**Reply by betterRunner** [reply]
了解，目前我通过设置以下代码实现了：

```
registerValidateLocale({
    'zh-CN': {
        required: '请输入<% text %>'
    }
});

registerValidateMessageTemplateEnigne((message, context) => {
    return message.replace(/\<\%\s*([\w.]+)\s*\%\>/g, (_, $0) => {
        console.log('s0', $0, _, context)
        return FormPath.getIn(context, $0)
    });
});
```

schema部分：
```
'required': true,
'x-validator': {
    required: {
        text: true, // 设置text字段为true
    },
    text: 'your text' // 设置自定义文案
}
```

不知道有没有坑

**Reply by janryWang** [reply]
你这么说倒是提醒我了，你不用内置的模板引擎语法，是因为跟json-schema本身的表达式冲突了吧

**Reply by betterRunner** [reply]
对是的，现在2.x似乎不能对required规则进行message修改（以前1.x是可以的）。如果可以修改required的规则就不用这么麻烦了：
```
required: {
  message: '请输入xxx'
}
```

**Reply by janryWang** [reply]
要不你直接提个pr把内置语法改成<%吧，这个现在应该没人用到

**Comment by cgfeel** [comment]
> 了解，目前我通过设置以下代码实现了：
> 
> ```
> registerValidateLocale({
>     'zh-CN': {
>         required: '请输入<% text %>'
>     }
> });
> 
> registerValidateMessageTemplateEnigne((message, context) => {
>     return message.replace(/\<\%\s*([\w.]+)\s*\%\>/g, (_, $0) => {
>         console.log('s0', $0, _, context)
>         return FormPath.getIn(context, $0)
>     });
> });
> ```
> 
> schema部分：
> 
> ```
> 'required': true,
> 'x-validator': {
>     required: {
>         text: true, // 设置text字段为true
>     },
>     text: 'your text' // 设置自定义文案
> }
> ```
> 
> 不知道有没有坑


谢，搜索找到答案了 @betterRunner
